Refunds endpoint implementation

0nko opened the pull request:
Fixes #1308. 

This PR adds API endpoints for creating, fetching & getting all refunds of an order, or specific individual refunds. What's specific about this implementation is that coroutines are used here, similar to WPAndroid Stats.

There are RefundStore unit tests and code from this PR is used by woocommerce/woocommerce-android#1410, where its functionality can be tested, as well.

anitaa1990 commented on fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/jetpacktunnel/JetpackTunnelGsonRequestBuilder.kt:
```diff
@@ -0,0 +1,125 @@
+     * @param params the parameters to append to the request URL
+     * @param clazz the class defining the expected response
+     */
+    suspend fun <T : Any> syncGetRequest(
```
np: The comments for this method is the same as the comment for the method defined above. Could it be worded differently to explain how this method differs from the above one?

anitaa1990 commented on fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/jetpacktunnel/JetpackTunnelGsonRequestBuilder.kt:
```diff
@@ -0,0 +1,125 @@
+     * @param body the content body, which will be converted to JSON using [Gson][com.google.gson.Gson]
+     * @param clazz the class defining the expected response
+     */
+    suspend fun <T : Any> syncPostRequest(
```
Nice!! 👍

anitaa1990 commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/persistence/RefundsSqlUtils.kt:
```diff
@@ -0,0 +1,89 @@
+import org.wordpress.android.fluxc.model.SiteModel
+import org.wordpress.android.fluxc.network.rest.wpcom.wc.refunds.RefundsRestClient.RefundResponse
+
+object RefundsSqlUtils {
```
np: it would be nice to have a test class to test insert, update, delete for this class. Some reference classes: `ProductSqlUtilsTest`, `WCStatsSqlUtilsTest`

anitaa1990 commented on example/src/main/res/layout/fragment_woo_refunds.xml:
```diff
@@ -0,0 +1,65 @@
+            android:layout_margin="20dp"
+            android:background="@android:color/darker_gray" />
+
+        <EditText
```
np: Do you think it makes more sense to create a reusable `EditText` component here, rather than creating a new `EditText` for each button? I believe there is an implementation [here](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/dd1c416ba70bea7bdf27682c7e73c44f61f6d3a0/example/src/main/java/org/wordpress/android/fluxc/example/utils/AlertDialogUtils.kt#L9) to display an `EditText` dialog that can be reused.

0nko replied:
Oh nice, this makes it much easier to build the testing UI! Thanks! :)

anitaa1990 commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/RefundsStore.kt:
```diff
@@ -0,0 +1,140 @@
+        amount: BigDecimal,
+        reason: String = "",
+        autoRefund: Boolean = false
+    ) =
```
np: Completely non blocking but I _think_ our convention for functions with inferred return types is that they should be one or two lines.

anitaa1990 commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/RefundsStore.kt:
```diff
@@ -0,0 +1,140 @@
+    }
+
+    suspend fun fetchRefund(site: SiteModel, orderId: Long, refundId: Long) =
+        withContext(coroutineContext) {
```
np: Same comment as above. Completely non blocking but I _think_ our convention for functions with inferred return types is that they should be one or two lines.

0nko commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/model/refunds/RefundModel.kt:
```diff
@@ -0,0 +1,26 @@
+import java.math.BigDecimal
+import java.util.Date
+
+data class RefundModel(
```
Yeah, I saw that and I don't like it :P. It's like the NS prefix in every framework class in iOS. It's just noise. It doesn't help with anything, there is no refund or order entity outside of WooCommerce, so there is no need for the extra acronym. Plus, even if there were, they'd be in a different namespace/package. 

I understand that for consistency they should all follow the same scheme. In fact, the WordPress models don't have the WP prefix, either (`PageModel`, `PlanModel`, `AccountModel`, etc.). Should we rename the other WC models or do you just want to keep it as is?

cc: @AmandaRiu, @nbradbury

aforcier replied:
A bit of background on the `WC` prefix:

The idea with FluxC was that it can eventually support any number of plugins, each of them a separate gradle module like WooCommerce is. This would mean that while there currently isn't a refund or order entity outside of WooCommerce, there might be in the future from a separate plugin not yet supported.

Package namespace is of course helpful in keeping the compiler happy, but it doesn't entirely protect us since we usually use the IDE to auto-import - so it's relatively easy to import the wrong `OrderModel` and not notice, for example. And in fact it a lot of cases it will still cause problems for the compiler, since the package for `RefundModel` is actually `org.wordpress.android.fluxc.model.refunds`, so it will conflict with another plugin that used the `model.refunds` when both plugins are imported. Many of our other models and classes are right in the root package (e.g. `model`, or `store`) so it's even easier for issues to arise for those.

The reason the WP ones don't have the prefix is because that's part of 'core' FluxC. Perhaps this will turn out to have been short-sighted, but at the moment you can't even use core FluxC functionality without importing those classes (so any FluxC plugin submodule is importing those classes and will catch a conflict early). So the onus would be on the developer supporting a new plugin to use naming prefixes.

I agree that having to add `WC` is pretty clunky, but if we opt not to use it we should have an alternate way of saving ourselves from future confusion and refactors. Do you see an alternate way to do this that dodges the issues I mentioned?

0nko replied:
Thanks for the clarification, @aforcier. I'm just curious, was there a reason why the Woo plugin doesn't use a unique package name that would avoid a potential conflict with other plugins? Anyway, your explanation makes sense and now that I know there's a reason behind the naming I don't have a problem with it.

aforcier replied:
Good question - IIRC I considered a unique package name and there were a few blockers, so I ended up going with this approach. I don't recall all the reasons, but from what I remember:

Some of our (Java) base classes are package private to constrain their use outside of FluxC - therefore, things like `WooCommerceStore` would need to be declared in the same package to be able to extend `Store`.

I believe there were also some issues with the tests that were resolvable by using the same package. And probably our code generator utilities also relied on things to belong to a certain package for discoverability.

All of those were probably addressable but together were big enough blockers to stop considering it and opt for sharing the package. And as I mentioned I wasn't convinced that separate namespacing was even necessarily the best approach since it's pretty easy to auto-import the wrong class when many share the same name.

AmandaRiu replied:
I'd say since we are already using this prefix in FluxC (I'm very used to always typing `WC` to get a list of models specific to Woo) we should probably keep it that way. It also differentiates the FluxC-related models from the client models. I've been creating smaller more manageable models in the client code and they don't start with `WC`.

0nko replied:
Sounds good. Thanks for the answer, Alex.

0nko commented on example/src/main/java/org/wordpress/android/fluxc/example/ui/refunds/WooRefundsFragment.kt:
```diff
@@ -0,0 +1,107 @@
+                GlobalScope.launch {
+                    val response = refundsStore.fetchRefund(
+                            site,
+                            order_number.text.toString().toLong(),
```
The UI before was bad. I wrote it to test the code quickly but it would crash if it didn't get all the parameters it required. I fixed it and you shouldn't see the crash now.

0nko commented:
@anitaa1990 Thank you for the thorough review & testing! I've addressed all the issues you've mentioned and there shouldn't be any more crashes in the example app 🤞.

Something I've noticed, though -- one of the `DateUtils` unit tests (testGetStartDayOfCurrentMonthForSite) is failing, but it's not from any of the changes I've made. Do you know what could be causing it?

anitaa1990 commented:
The changes looks good @0nko! I was taking a look at the failing unit test and it seems to be because of a timezone difference. I modified the test and committed the changes here so that we can merge to this PR. I hope that is ok. 

There is a small 
lint error and lint warning in the `RefundsStoreTest` that needs to be fixed 

#### Lint error:
```
Exceeded max line length (120) (cannot be auto-corrected)
```

#### Lint warning:
 ```
`Warning:(54, 17) This declaration is experimental and its usage should be marked with '@kotlinx.coroutines.ExperimentalCoroutinesApi' or '@UseExperimental(kotlinx.coroutines.ExperimentalCoroutinesApi::class)'
```

Let me know once that is done and we can merge 👍

0nko commented:
I fixed the lint error but I'm not seeing any warning 🤷‍♂.
